util/keydecoder/keydecoder.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keydecoder
+
+import (
+	"fmt"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// HandleType is the type of a Handle, could be `int`(ie. kv.IntHandle) or `common`(ie. *kv.CommonHandle)
+type HandleType string
+
+// HandleType instances
+const (
+	IntHandle    HandleType = "int"
+	CommonHandle HandleType = "common"
+
+	UnknownHandle HandleType = "unknown"
+)
+
+// DecodedKey is a struct contains detailed information about a key, its json form should be used to fill KEY_INFO field in `DEADLOCKS` and `DATA_LOCK_WAITS`
+type DecodedKey struct {
+	DbID              int64      `json:"db_id,omitempty"`
+	DbName            string     `json:"db_name,omitempty"`
+	TableID           int64      `json:"table_id"`
+	TableName         string     `json:"table_name,omitempty"`
+	HandleType        HandleType `json:"handle_type,omitempty"`
+	IsPartitionHandle bool       `json:"partition_handle,omitempty"`
+	HandleValue       string     `json:"handle_value,omitempty"`
+	IndexID           int64      `json:"index_id,omitempty"`
+	IndexName         string     `json:"index_name,omitempty"`
+	IndexValues       []string   `json:"index_values,omitempty"`
+}
+
+func handleType(handle kv.Handle) HandleType {
+	if _, ok := handle.(kv.IntHandle); ok {
+		return IntHandle
+	} else if _, ok := handle.(*kv.CommonHandle); ok {
+		return CommonHandle
+	} else if h, ok := handle.(kv.PartitionHandle); ok {
+		return handleType(h.Handle)
+	} else if h, ok := handle.(*kv.PartitionHandle); ok {
+		return handleType(h.Handle)
+	} else {
+		logutil.BgLogger().Warn("Unexpected kv.Handle type",
+			zap.Any("handle", handle),
+			zap.String("handle Type", fmt.Sprintf("%T", handle)),
+		)
+	}
+	return UnknownHandle
+}
+
+// DecodeKey decodes `key` into `DecodedKey`, which is used to fill KEY_INFO field in `DEADLOCKS` and `DATA_LOCK_WAITS`
```

Comment:
May need to make it clear that only user data key and index key could be used as the input parameter, other key types should not be used like the meta keys. Better to add some defensive logic to  verify if the key type is expected or report error if the input is unexpected.

Reply:
I'm actually afraid that it's not so easy to let the user know what error happened, and the user only sees a "NULL" in this json column. We should not make the whole query fail when a key is unable to be decoded.